util/misc.py

```diff
@@ -410,4 +412,106 @@ def which(application_binary_name):
         full_path=os.path.join(path, application_binary_name)
         if os.path.exists(full_path) and os.access(full_path, os.X_OK):
             link_resolved_path = os.path.realpath(full_path)
-            return link_resolved_path
\ No newline at end of file
+            return link_resolved_path
+
+def is_nonstr_seq(x):
+    '''Tests whether `x` is a Sequence other than a string'''
+    return isinstance(x, collections.Sequence) and not isinstance(x,(str,bytes))
+
+def make_seq(x):
+    '''Return a tuple containing the items in `x`, or containig just `x` if `x` is a non-string sequence.  Convenient
+    for uniformly writing iterations over parameters that may be passed in as an item or a tuple/list of items.
+    '''
+    return tuple(x) if is_nonstr_seq(x) else (x,)
+
+def flatten_dict(d):
+    '''Return a new dict `r`, where r[(k1,k2,...,kn)]==v iff d[k1][k2]...[kn]==v.'''
+    result=dict()
+    for k, v in d.items():
+        assert not isinstance(k,tuple), 'Dict already flattened? {}'.format(d)
+        if isinstance(v, collections.Mapping):
+            for k_rest, v_actual in flatten_dict(v).items():
+                result[(k,)+k_rest]=v_actual
+        else:
+            result[(k,)]=v
+    return result
+#    return dict(((k,)+k1,v1) for k, v in d.items() for k1, v1 in \
+#                 (flatten_dict(v).items() if isinstance(v, collections.Mapping) else [((),v)]))
+
+def unflatten_dict(d):
+    '''Reverse operation to `flatten_dict`'''
+    result = dict()
+    for k, v in d.items():
+        # if k is (k1,k2,...kN), then ensure that result[k1][k2]...[kN]=v, creating intermediate dicts as needed
+        functools.reduce(lambda d,k: d.setdefault(k,dict()), k[:-1], result)[k[-1]]=v
+    return result
+
+def load_yaml_or_json(fname):
+    '''Load a dictionary from either a yaml or a json file'''
+    with open(fname) as f:
+        if fname.upper().endswith('.YAML'): return yaml.safe_load(f)
+        if fname.upper().endswith('.JSON'): return json.load(f)
+        raise TypeError('Unsupported dict file format: ' + fname)
+
+def load_config(cfg, include_directive='include', std_includes=[], param_renamings={}):
+    '''Load a config file, recursively loading any included config files; load any params specified under legacy names.
+
+    This function implements extensions to the standard way of specifying configuration parameters via (possibly nested)
+    dictionaries.  The extensions let config files include other config files, and let us rename parameters while
+    still letting them be specified under old names for backwards compatibility.
```

Comment:
I think this needs more description. Reading the code there are two things undocumented.

* Included filenames can be the filename without path if they are in the same directory as cfg AND cfg is a filename.
* Param renaming renames parameters in cfg.
* Order of include overrides is std_includes, include_directive, then finally cfg.